### DISCUSSION TITLE
SFTP Source Reader Javascript - Source transformer

### DISCUSSION THREAD

**Discussion by Ruffdraft** [nextgenhealthcare/connect#4584] [Q&A]
Anyone know if the function can be called from a Source transformer?
I'm Looking to use it as a SFTP reader and read files in subfolders. But its not doing anything, no errors or logs from what I can tell.

Code:

`sftpGet("127.0.0.1",22,"username","password",true);`

I've modified the 'called' function itself for my env,
Added the function to Code Templates,
Set dependencies in code Template lib,
Set Source type set to Channel Reader.

thanks. 

```
function sftpGet(ip,port,user,pass,remove){
	importPackage(com.jcraft.jsch);
	importPackage(java.io);
	importPackage(java.lang);
	importPackage(java.nio.charset);
	
	function toByteArray(str){
		var bytes = [];
		for (var i = 0; i < str.length; ++i) {
    		bytes.push(str.charCodeAt(i));
		}
		return bytes;
	}
	
	// testing vars
	var verbose = false;
	
	//create object.
	var jsch = new JSch();
	jsch.setConfig('StrictHostKeyChecking','no');
	var session = jsch.getSession(user,ip,port);
	session.setPassword(pass);
	session.setTimeout(2000);
	session.connect();
	var channel = session.openChannel('sftp');
	var ioexception = new Packages.java.io.IOException;
	channel.connect();
	if(verbose){logger.info('Connected to SFTP: ');}
	if(verbose){logger.info('Home Directory: ' + channel.getHome());}
	channel.cd('/'); //Folder on the SFTP server where file will be found
	var counter
	var directoryList = channel.ls("*.txt");
	if(directoryList.size()>0||verbose){logger.info('Number of Files on Server ' + directoryList.size());}
	if (directoryList.size() > 0){
		for(var i=0; i<directoryList.size();i++) {
			var filename = directoryList.get(i).getFilename();
			logger.info('File Name: ' + filename);
			var is = channel.get(filename);
			var bReader = new BufferedReader(new InputStreamReader(is));
			var sbfFileContents = "";
			var line;
			//read file line by line
	        while((line = bReader.readLine())!=null&&line.length()!=0){
        		sbfFileContents+=line;
				}
			var stream = new ByteArrayInputStream(toByteArray(sbfFileContents));
	/*******************/
	channel.put(stream,"test/"+filename); //copies file to subfolder for testing
	/*******************/

			// Remove After Processing
			if(remove){channel.rm(filename);}
		}   
	}
	//Close everything down
	channel.disconnect();
	session.disconnect();	
	return;
}
```

**Comment by rjkroll** [comment]
How are you kicking off the channel if its a Channel Reader?   Do you have the 'test' folder already created that the code is trying to write the file to?

**Comment by jonbartels** [comment]
> Set Source type set to Channel Reader.

Change your source to a Javascript Reader. This type is a polling type and literally runs Javascript to generate the source message. Once the connector polls and generates a source message then it will kick off any source transformers as usual. A dummy JS Reader would just call `return "Look at meeee! I am a message!"` and then `msg` would be `Look at meeee! I am a message!`. Of course you can return JSON or HL7 or XML here too, but I think for your use case you just want to return a "go now" kind of message to trigger your SFTP reading function.

It *may* make sense to call your sftp function from the JS Reader directly instead of as a source transformer, however I do not understand the use case well enough to make a suggestion about t hat.
